keras/layers/core.py

```diff
@@ -471,59 +516,65 @@ def get_output(self, train=False):
     '''
-    def __init__(self, input_dim, output_dim, init='glorot_uniform', activation='linear', weights=None, name=None,
-                 W_regularizer=None, b_regularizer=None, activity_regularizer=None,
-                 W_constraint=None, b_constraint=None):
+    input_ndim = 2
 
-        super(Dense, self).__init__()
+    def __init__(self, output_dim, init='glorot_uniform', activation='linear', weights=None,
+                 W_regularizer=None, b_regularizer=None, activity_regularizer=None,
+                 W_constraint=None, b_constraint=None, input_dim=None, **kwargs):
         self.init = initializations.get(init)
         self.activation = activations.get(activation)
-        self.input_dim = input_dim
         self.output_dim = output_dim
 
+        self.W_regularizer = regularizers.get(W_regularizer)
+        self.b_regularizer = regularizers.get(b_regularizer)
+        self.activity_regularizer = regularizers.get(activity_regularizer)
+
+        self.W_constraint = constraints.get(W_constraint)
+        self.b_constraint = constraints.get(b_constraint)
+        self.constraints = [self.W_constraint, self.b_constraint]
+
+        self.initial_weights = weights
+
+        self.input_dim = input_dim
+        if self.input_dim:
+            kwargs['input_shape'] = (self.input_dim,)
+        super(Dense, self).__init__(**kwargs)
+
+    def build(self):
+        input_dim = self.input_shape[1]
+
         self.input = T.matrix()
-        self.W = self.init((self.input_dim, self.output_dim))
-        self.b = shared_zeros((self.output_dim))
+        self.W = self.init((input_dim, self.output_dim))
+        self.b = shared_zeros((self.output_dim,))
 
         self.params = [self.W, self.b]
 
         self.regularizers = []
-        self.W_regularizer = regularizers.get(W_regularizer)
         if self.W_regularizer:
             self.W_regularizer.set_param(self.W)
             self.regularizers.append(self.W_regularizer)
 
-        self.b_regularizer = regularizers.get(b_regularizer)
         if self.b_regularizer:
             self.b_regularizer.set_param(self.b)
             self.regularizers.append(self.b_regularizer)
 
-        self.activity_regularizer = regularizers.get(activity_regularizer)
         if self.activity_regularizer:
             self.activity_regularizer.set_layer(self)
             self.regularizers.append(self.activity_regularizer)
 
-        self.W_constraint = constraints.get(W_constraint)
-        self.b_constraint = constraints.get(b_constraint)
-        self.constraints = [self.W_constraint, self.b_constraint]
-
-        if weights is not None:
-            self.set_weights(weights)
-
-        if name is not None:
-            self.set_name(name)
-
-    def set_name(self, name):
-        self.W.name = '%s_W' % name
-        self.b.name = '%s_b' % name
+        if self.initial_weights is not None:
+            self.set_weights(self.initial_weights)
+            del self.initial_weights
```

Comment:
The weight setting boilerplate code is repeated in all layers, would there be a way to abstract it?